frame/evm/precompile/modexp/src/lib.rs

```diff
@@ -34,44 +34,52 @@ const MIN_GAS_COST: u64 = 200;
 
 // Calculate gas cost according to EIP 2565:
 // https://eips.ethereum.org/EIPS/eip-2565
-fn calculate_gas_cost(
-	base_length: u64,
-	exp_length: u64,
-	mod_length: u64,
-	exponent: &BigUint,
-) -> u64 {
+fn calculate_gas_cost(base_length: u64, mod_length: u64, exponent: &BigUint) -> u64 {
 	fn calculate_multiplication_complexity(base_length: u64, mod_length: u64) -> u64 {
 		let max_length = max(base_length, mod_length);
 		let mut words = max_length / 8;
 		if max_length % 8 > 0 {
 			words += 1;
 		}
 
-		// TODO: prevent/handle overflow
+		// Note: can't overflow because we take words to be some u64 value / 8, which is
+		// necessarily less than sqrt(u64::MAX).
+		// Additionally, both base_length and mod_length are bounded to 1024, so this has
+		// an upper bound of roughly (1024 / 8) squared
 		words * words
 	}
 
-	fn calculate_iteration_count(exp_length: u64, exponent: &BigUint) -> u64 {
+	fn calculate_iteration_count(exponent: &BigUint) -> u64 {
+		let exp_length_bytes = (exponent.bits() + 7) / 8;
 		let mut iteration_count: u64 = 0;
 
-		if exp_length <= 32 && exponent.is_zero() {
+		if exponent.is_zero() {
 			iteration_count = 0;
-		} else if exp_length <= 32 {
+		} else if exp_length_bytes <= 32 {
 			iteration_count = exponent.bits() - 1;
-		} else if exp_length > 32 {
+		} else if exp_length_bytes > 32 {
 			// construct BigUint to represent (2^256) - 1
 			let bytes: [u8; 32] = [0xFF; 32];
 			let max_256_bit_uint = BigUint::from_bytes_be(&bytes);
 
+			// from the EIP spec:
+			// (8 * (exponent_length - 32)) + ((exponent & (2**256 - 1)).bit_length() - 1)
+			//
+			// Notes:
+			// * exp_length_bytes is bounded to 1024
+			// * exponent can't be 0 because we ensure exp_length_bytes is > 32
+			// * exponent.bitand(max_256_bit_uint) can't be 0, so subtracting one can't underflow
+			// * the addition can't overflow because the terms are both capped at roughly
+			//   8 * max size of exp_length_bytes (1024)
 			iteration_count =
-				(8 * (exp_length - 32)) + ((exponent.bitand(max_256_bit_uint)).bits() - 1);
+				(8 * (exp_length_bytes - 32)) + ((exponent.bitand(max_256_bit_uint)).bits() - 1);
```

Comment:
```suggestion
				(8 * (exp_length - 32)) + (exponent.bitand(max_256_bit_uint)).bits() - 1;
```

Reply:
Fixed in 8c8370a (had to remove two sets of parens)